File: examples/node/heartbeat.mjs
#!/usr/bin/env node
// GotYouBro heartbeat client for Node.js 18+ (no dependencies).
//
//   GOTYOUBRO_URL=https://gotyoubro.example.com GOTYOUBRO_TOKEN=gyb_... node heartbeat.mjs
//   node heartbeat.mjs --once          # single heartbeat, e.g. at the end of a cron job
//   GOTYOUBRO_MONITOR=worker node heartbeat.mjs   # heartbeat for the service's "worker" monitor
//
// To use it inside your app, copy `sendHeartbeat` and call `startHeartbeat()` once at startup.

import { pathToFileURL } from 'node:url';

const BASE_URL = (process.env.GOTYOUBRO_URL ?? 'http://localhost:6969').replace(/\/+$/, '');
const TOKEN = process.env.GOTYOUBRO_TOKEN;
const INTERVAL_SECONDS = Number(process.env.GOTYOUBRO_INTERVAL ?? 30);
// Optional monitor key (a service can have several monitors: api, worker, nightly-job…).
const MONITOR = process.env.GOTYOUBRO_MONITOR ?? '';

export async function sendHeartbeat() {
  const response = await fetch(`${BASE_URL}/api/v1/health/heartbeat${MONITOR ? `/${encodeURIComponent(MONITOR)}` : ''}`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${TOKEN}` },
    signal: AbortSignal.timeout(10_000),
  });
  const body = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new Error(`${response.status} ${body.error?.code ?? ''} ${body.error?.message ?? ''}`.trim());
  }
  return body.data; // { serviceId, healthStatus, lastHeartbeatAt, nextHeartbeatDeadline, recovered }
}

/** Send a heartbeat now and then every `intervalSeconds`. Errors are logged, never thrown. */
export function startHeartbeat(intervalSeconds = INTERVAL_SECONDS) {
  const beat = async () => {
    try {
      const data = await sendHeartbeat();
      console.log(`[heartbeat] ${new Date().toISOString()} ${data.healthStatus}${data.recovered ? ' (recovered)' : ''}`);
    } catch (err) {
      // Monitoring must never take your app down.
      console.error(`[heartbeat] failed: ${err.message}`);
    }
  };
  beat();
  const timer = setInterval(beat, intervalSeconds * 1000);
  timer.unref?.(); // don't keep the process alive just for heartbeats
  return () => clearInterval(timer);
}

// Run as a script.
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  if (!TOKEN) {
    console.error('Set GOTYOUBRO_TOKEN (and GOTYOUBRO_URL).');
    process.exit(2);
  }
  if (process.argv.includes('--once')) {
    sendHeartbeat()
      .then((data) => console.log(`heartbeat ok: ${data.healthStatus}`))
      .catch((err) => {
        console.error(`heartbeat failed: ${err.message}`);
        process.exit(1);
      });
  } else {
    startHeartbeat();
    setInterval(() => {}, 1 << 30); // keep the standalone script running
  }
}

File: examples/python/heartbeat.py
#!/usr/bin/env python3
"""GotYouBro heartbeat client for Python 3.8+ (standard library only).

    GOTYOUBRO_URL=https://gotyoubro.example.com GOTYOUBRO_TOKEN=gyb_... python3 heartbeat.py
    python3 heartbeat.py --once        # single heartbeat, e.g. at the end of a cron job
    GOTYOUBRO_MONITOR=worker python3 heartbeat.py   # heartbeat for the "worker" monitor

To use it inside your app, import `start_heartbeat` and call it once at startup; it runs in a
daemon thread and never raises.
"""

import json
import os
import sys
import threading
import time
import urllib.error
import urllib.parse
import urllib.request

BASE_URL = os.environ.get("GOTYOUBRO_URL", "http://localhost:6969").rstrip("/")
TOKEN = os.environ.get("GOTYOUBRO_TOKEN", "")
INTERVAL_SECONDS = float(os.environ.get("GOTYOUBRO_INTERVAL", "30"))
# Optional monitor key (a service can have several monitors: api, worker, nightly-job...).
MONITOR = os.environ.get("GOTYOUBRO_MONITOR", "")


def send_heartbeat() -> dict:
    """Send one heartbeat. Returns the response data; raises on failure."""
    request = urllib.request.Request(
        f"{BASE_URL}/api/v1/health/heartbeat" + (f"/{urllib.parse.quote(MONITOR)}" if MONITOR else ""),
        method="POST",
        headers={"Authorization": f"Bearer {TOKEN}"},
    )
    try:
        with urllib.request.urlopen(request, timeout=10) as response:
            return json.load(response)["data"]
    except urllib.error.HTTPError as err:
        try:
            error = json.load(err).get("error", {})
        except ValueError:
            error = {}
        raise RuntimeError(f"{err.code} {error.get('code', '')} {error.get('message', '')}".strip()) from None


def start_heartbeat(interval_seconds: float = INTERVAL_SECONDS) -> threading.Event:
    """Send heartbeats in a background thread. Set the returned event to stop."""
    stop = threading.Event()

    def loop() -> None:
        while not stop.is_set():
            try:
                data = send_heartbeat()
                suffix = " (recovered)" if data.get("recovered") else ""
                print(f"[heartbeat] {time.strftime('%Y-%m-%dT%H:%M:%S')} {data['healthStatus']}{suffix}", flush=True)
            except Exception as exc:  # monitoring must never take your app down
                print(f"[heartbeat] failed: {exc}", file=sys.stderr, flush=True)
            stop.wait(interval_seconds)

    threading.Thread(target=loop, name="gotyoubro-heartbeat", daemon=True).start()
    return stop


if __name__ == "__main__":
    if not TOKEN:
        sys.exit("Set GOTYOUBRO_TOKEN (and GOTYOUBRO_URL).")
    if "--once" in sys.argv:
        try:
            print(f"heartbeat ok: {send_heartbeat()['healthStatus']}")
        except Exception as exc:
            sys.exit(f"heartbeat failed: {exc}")
    else:
        start_heartbeat()
        try:
            while True:
                time.sleep(3600)
        except KeyboardInterrupt:
            pass
